File: 섹션08. A형 문제 풀이 (~2022 상반기)/12. 연산자 배치하기 (코드트리, 2017 하반기 오후 2번).cpp
#include <stdio.h>

#define MAX (11 + 5)
#define INF (0x7fff0000)

#define PLUS (0)
#define MINUS (1)
#define MULTIPLY (2)

int T;

int N;
int num_of_cases[MAX];
int visit[MAX];
int possible[3]; // 사용할 수 있는 최대 PLUS, MINUS, MULTIPLY 
int number[MAX];

int minAnswer, maxAnswer;

void input()
{
	scanf("%d", &N);

	for (int i = 0; i < N; i++) scanf("%d", &number[i]);
	for (int i = 0; i < 3; i++) scanf("%d", &possible[i]);
}

void printCases()
{
	for (int i = 0; i < N - 1; i++) printf("%d ", num_of_cases[i]);
	putchar('\n');
}

int calculate()
{
	int sum = number[0];
	for (int i = 0; i < N - 1; i++)
	{
		if (num_of_cases[i] == PLUS) sum += number[i + 1];
		else if (num_of_cases[i] == MINUS) sum -= number[i + 1];
		else if (num_of_cases[i] == MULTIPLY) sum *= number[i + 1];
	}

	return sum;
}

void DFS(int depth)
{
	if (depth == N - 1)
	{
		// printCases();
		
		int tmp = calculate();
		if (maxAnswer < tmp) maxAnswer = tmp;
		if (minAnswer > tmp) minAnswer = tmp;

		return;
	}

	for (int i = 0; i < 3; i++)
	{
		if (visit[i] == possible[i]) continue;

		num_of_cases[depth] = i;

		visit[i]++;
		DFS(depth + 1);
		visit[i]--;
	}
}

int main()
{
	// scanf("%d", &T);
	T = 1;
	for (int tc = 1; tc <= T; tc++)
	{
		input();

		minAnswer = INF;
		maxAnswer = -INF;

		DFS(0);

		printf("%d %d\n", minAnswer, maxAnswer);
	}

	return 0;
}

File: Python/섹션08. A형 문제 풀이 (~2022 상반기)/12. 연산자 배치하기 (코드트리, 2017 하반기 오후 2번).py
MAX = 11 + 5
INF = 0x7fff0000

PLUS = 0
MINUS = 1
MULTIPLY = 2

num_of_cases = [0] * MAX
visit = [0] * MAX
possible = [0] * 3 # 사용할 수 있는 최대 PLUS, MINUS, MULTIPLY 
number = [0] * MAX

def input_data():
    global N, number, possible
    
    N = int(input())
    
    number = list(map(int, input().split()))
    possible = list(map(int, input().split()))

def print_cases():
    print(*print_cases[:N - 1])

def calculate():
    sum = number[0]
    for i in range(N - 1):
        if num_of_cases[i] == PLUS: sum += number[i + 1]
        elif num_of_cases[i] == MINUS: sum -= number[i + 1]
        elif num_of_cases[i] == MULTIPLY: sum *= number[i + 1]
        
    return sum

def DFS(depth):
    global min_answer, max_answer
    
    if depth == N - 1:
        # print_cases()
        
        tmp = calculate()
        if max_answer < tmp: max_answer = tmp
        if min_answer > tmp: min_answer = tmp
            
        return

    for i in range(3):
        if visit[i] == possible[i]: continue

        num_of_cases[depth] = i
        
        visit[i] += 1
        DFS(depth + 1)
        visit[i] -= 1

# T = int(input())
T = 1
for tc in range(T):
    global min_answer, max_answer
    
    input_data()
    
    min_answer = INF
    max_answer = -INF
    
    DFS(0)
    
    print(min_answer, max_answer)